server/src/main/java/org/elasticsearch/cluster/metadata/TemplateUpgradeService.java

```diff
@@ -175,27 +172,48 @@ public void onFailure(Exception e) {
             client.admin().indices().deleteTemplate(request, new ActionListener<DeleteIndexTemplateResponse>() {
                 @Override
                 public void onResponse(DeleteIndexTemplateResponse response) {
-                    updatesInProgress.decrementAndGet();
                     if (response.isAcknowledged() == false) {
+                        allUpgradesSuccessful = false;
                         logger.warn("Error deleting template [{}], request was not acknowledged", template);
                     }
+                    tryFinishUpgrade();
                 }
 
                 @Override
                 public void onFailure(Exception e) {
-                    updatesInProgress.decrementAndGet();
+                    allUpgradesSuccessful = false;
                     if (e instanceof IndexTemplateMissingException == false) {
                         // we might attempt to delete the same template from different nodes - so that's ok if template doesn't exist
                         // otherwise we need to warn
                         logger.warn(new ParameterizedMessage("Error deleting template [{}]", template), e);
                     }
+                    tryFinishUpgrade();
                 }
             });
         }
     }
 
-    int getUpdatesInProgress() {
-        return updatesInProgress.get();
+    void tryFinishUpgrade() {
+        assert upgradesInProgress.get() > 0;
+        if (upgradesInProgress.decrementAndGet() == 1) {
+            // this is the last upgrade, the templates should now be in the desired state
+            if (allUpgradesSuccessful) {
+                logger.info("Templates were upgraded successfuly to version {}", Version.CURRENT);
+            } else {
+                logger.info("Templates were partially upgraded to version {}", Version.CURRENT);
+            }
+            allUpgradesSuccessful = true;
+            // Check upgraders are satisfied after the update completed. If they still
+            // report that changes are required, this might indicate a bug or that something
+            // else tinkering with the templates during the upgrade.
+            final ImmutableOpenMap<String, IndexTemplateMetaData> upgradedTemplates = clusterService.state().getMetaData().getTemplates();
+            final boolean changesRequired = calculateTemplateChanges(upgradedTemplates).isPresent();
+            if (changesRequired) {
+                logger.warn("Templates are still reported as out of date after the upgrade. The template upgrade will be retried.");
+            }
+            final int noMoreUpgrades = upgradesInProgress.decrementAndGet();
```

Comment:
I think we should do this as one of the first steps in the if block. The reasoning is that if an exception occurs before this, we will never get the chance to continue operating as nothing will trigger this method again

Reply:
good point! I have enclosed the last sanity check inside a try-finally block, and the finally branch will make sure the counter is decremented. The decrement has to stay at the end as its purpose is to guard the check.